compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/ir/IrUtils.kt

```diff
@@ -319,36 +313,23 @@ fun IrType.remapTypeParameters(
 ): IrType =
     when (this) {
         is IrSimpleType -> {
-            val classifier = classifier.owner
-            when {
-                classifier is IrTypeParameter -> {
-                    val newClassifier =
-                        srcToDstParameterMap?.get(classifier) ?:
-                        if (classifier.parent == source)
-                            target.typeParameters[classifier.index]
-                        else
-                            classifier
-                    IrSimpleTypeImpl(newClassifier.symbol, hasQuestionMark, arguments, annotations)
-                }
-
-                classifier is IrClass ->
-                    IrSimpleTypeImpl(
-                        classifier.symbol,
-                        hasQuestionMark,
-                        arguments.map {
-                            when (it) {
-                                is IrTypeProjection -> makeTypeProjection(
-                                    it.type.remapTypeParameters(source, target, srcToDstParameterMap),
-                                    it.variance
-                                )
-                                else -> it
-                            }
-                        },
-                        annotations
+            val oldClassifier = classifier
+            val newClassifier = when {
+                oldClassifier is IrTypeParameterSymbol && oldClassifier.owner.parent == source ->
```

Comment:
This condition is wrong; see the test `typeParametersInLocalFunction.kt`, where type parameter `C` is captured by `inner2()`, but will not be remapped by your condition.
(Unfortunately, this does not lead to a visible test failure).

Reply:
I'll revert this change for now, this code should probably be refactored separately since such errors are not clearly visible.